### DISCUSSION TITLE
Anyone successfully running on M1 Mac?

### DISCUSSION THREAD

**Discussion by smozely** [batect/batect#963] [Q&A] [answered]
I'm trying to run my project on a M1 Mac, which successfully runs on Intel (my other machine)

But getting errors with FFI
`java.lang.UnsatisfiedLinkError: could not load FFI provider jnr.ffi.provider.jffi.Provider`

Which in a long stack trace, seems to be caused by `no matching architecture in universal wrapper`

So before I dig too much further, wonder if there are reports of others successfully running on M1? ... was kinda thinking I couldn't be the first

**Comment by charleskorn** [comment]
I don't have a M1 Mac to test, so I can't confirm this, but it _should_ work. What output do you get if you run `./batect --version`, `java -version` and `uname -m`?

**Reply by smozely** [reply]
Running `batect --version` ... gives a stacktrace ... 
```Fatal exception:
java.lang.UnsatisfiedLinkError: could not load FFI provider jnr.ffi.provider.jffi.Provider
        at jnr.ffi.provider.InvalidProvider$1.loadLibrary(InvalidProvider.java:49)
        at jnr.ffi.LibraryLoader.load(LibraryLoader.java:417)
        at jnr.posix.POSIXFactory$DefaultLibCProvider$SingletonHolder.<clinit>(POSIXFactory.java:307)
        at jnr.posix.POSIXFactory$DefaultLibCProvider.getLibC(POSIXFactory.java:335)
        at jnr.posix.BaseNativePOSIX.<init>(BaseNativePOSIX.java:41)
        at jnr.posix.MacOSPOSIX.<init>(MacOSPOSIX.java:16)
        at jnr.posix.POSIXFactory.loadMacOSPOSIX(POSIXFactory.java:153)
        at jnr.posix.POSIXFactory.loadNativePOSIX(POSIXFactory.java:121)
        at jnr.posix.POSIXFactory.getNativePOSIX(POSIXFactory.java:75)
        at jnr.posix.POSIXFactory.getNativePOSIX(POSIXFactory.java:85)
        at batect.ioc.BootstrapInjectionConfigurationKt$createKodeinConfiguration$1$2.invoke(BootstrapInjectionConfiguration.kt:47)
        at batect.ioc.BootstrapInjectionConfigurationKt$createKodeinConfiguration$1$2.invoke(BootstrapInjectionConfiguration.kt)
        at org.kodein.di.bindings.Singleton$getFactory$1$1$1.invoke(standardBindings.kt:134)
        at org.kodein.di.bindings.SingletonReference.make(references.kt:34)
```

Further down the stack is where this comes in .. 

```
Caused by: java.lang.UnsatisfiedLinkError: could not get native definition for type `POINTER`, original error message follows: java.lang.UnsatisfiedLinkError: Unable to execute or load jffi binary stub from `/var/folders/lj/wpydyxj96s9g8n4mylg1hkd40000gn/T/`. Set `TMPDIR` or Java property `java.io.tmpdir` to a read/write path that is not mounted "noexec".
/Users/steve/src/Koha/backend/jffi2619038268196344344.dylib: dlopen(/Users/steve/src/Koha/backend/jffi2619038268196344344.dylib, 1): no suitable image found.  Did find:
        /Users/steve/src/Koha/backend/jffi2619038268196344344.dylib: no matching architecture in universal wrapper
        /Users/steve/src/Koha/backend/jffi2619038268196344344.dylib: no matching architecture in universal wrapper
        at com.kenai.jffi.internal.StubLoader.tempLoadError(StubLoader.java:448)
        at com.kenai.jffi.internal.StubLoader.loadFromJar(StubLoader.java:433)
```

**Reply by smozely** [reply]
`uname -m` gives `arm64`

`java --version` gives `openjdk 15.0.3 2021-04-20`

... Though I've tried different java versions both Arm64 and x64 via Rosetta, with the same result. Though I'm not 100% sure I had the java setup correct, hence wanting to check if it had worked for anyone, before debugging further

**Reply by charleskorn** [reply]
OK, could you try running `java -XshowSettings:all 2>&1 | grep os.arch`? This will confirm what architecture the JVM is using.

The message in that stacktrace is also interesting - could you try setting the `TMPDIR` environment variable to something else like it suggests and see if that helps?

**Reply by smozely** [reply]
Here goes ...
```
java -XshowSettings:all 2>&1 | grep os.arch
    os.arch = aarch64
```
.. so yeah def arch64 version


With a different tmp dir set (this time set to a subfolder of the project), I get the same error message, just with the updated location ... I'm not sure where `noexec` is coming from, its just the standard MacOS root partition, I've done nothing interesting with my setup there.

**Reply by charleskorn** [reply]
It looks like there's something wrong with the architecture-specific library embedded inside Batect that is used to call some native functions. 

Could you please try these things?

* Are you able to check what version of Batect you're using? There should be a `VERSION="..."` line in your `batect` wrapper script in your project. If it's not 0.72.0, please try with that and let me know if that fixes the issue.
* If you're already using 0.72.0, try modifying your `batect` wrapper script like this around line 99:

    ```diff
                 ${GIT_BASH_PTY_WORKAROUND[@]+"${GIT_BASH_PTY_WORKAROUND[@]}"} \
                 java \
                 -Djava.net.useSystemProxies=true \
    +            -Djffi.extract.dir=$PWD/tmp \
    +            -Djffi.extract.name=thing.jnilib \
                 ${JAVA_OPTS[@]+"${JAVA_OPTS[@]}"} \
                 -jar "$JAR_PATH" \
                 "$@"
    ```
    
    This should extract the library to `./tmp/thing.jnilib.dylib`. Can you please double-check that the SHA256 checksum is correct? (Run `sha256sum ./tmp/thing.jnilib.dylib`.) It should be `b43a9219670551fb86c79bb4cf372162c8df4702bd38d7e07f5e72f068ce671b`.
    
    Running `file ./tmp/thing.jnilib.dylib` should give:
    
    ```
    ./tmp/thing.jnilib.dylib: Mach-O universal binary with 2 architectures: [x86_64:Mach-O 64-bit dynamically linked shared library x86_64] [arm64:Mach-O 64-bit dynamically linked shared library arm64]
    ./tmp/thing.jnilib.dylib (for architecture x86_64):	Mach-O 64-bit dynamically linked shared library x86_64
    ./tmp/thing.jnilib.dylib (for architecture arm64):	Mach-O 64-bit dynamically linked shared library arm64
    ```

**Reply by charleskorn** [reply]
This may actually be an issue in that library (`jffi` and `jnr-ffi`) - there's a similar report in their repo: https://github.com/jnr/jnr-ffi/issues/257#issuecomment-896079227

**Reply by smozely** [reply]
Same issue with those changes, good spotting on that JFFI issues, my searches didn't find that the other day. I guess we just follow that one and see where they get with it.

Thanks for the help!

**Reply by charleskorn** [reply]
@smozely, could you try running the following on your M1 Mac and then try to run Batect again? If so, I might be able to add a workaround for this issue until `jffi` fixes the issue for good:

```bash
cd ~/.batect/cache/0.74.0/ # Replace 0.74.0 with whatever version you're using
jar xf batect-0.74.0.jar jni/Darwin/libjffi-1.2.jnilib
codesign -s - ./jni/Darwin/libjffi-1.2.jnilib
jar uf batect-0.74.0.jar jni/Darwin/libjffi-1.2.jnilib
```

Note that this requires that you at least have the Xcode command line tools installed - if you have Xcode installed, or are using Homebrew, you should already have these available.

**Reply by smozely** [reply]
@charleskorn tried it, but no dice.

Couple of things extra I did
* set a BATECT_DOWNLOAD_CHECKSUM from the modified jar file
* remove those extra lines you suggested I add to the batect wrapper script ^ (which I should have removed already anyways)

But then when I run batect, it gets killed right away ... and in Crash Reports in my console app I see a crash report gets created with a lot of messages, but one important looking like "EXC_BAD_ACCESS (Code Signature Invalid)"

**Reply by charleskorn** [reply]
Interesting... if you go back to `~/.batect/cache/0.74.0/` and run `codesign -dv ./jni/Darwin/libjffi-1.2.jnilib`, what output do you get?

**Reply by smozely** [reply]
This is what I got ... 
```
Executable=/Users/steve/.batect/cache/0.74.0/jni/Darwin/libjffi-1.2.jnilib
Identifier=libjffi-1-55554944ddb5c3e598b5378987a6ec9eda8a36da
Format=Mach-O universal (x86_64 arm64)
CodeDirectory v=20400 size=1547 flags=0x2(adhoc) hashes=42+2 location=embedded
Signature=adhoc
Info.plist=not bound
TeamIdentifier=not set
Sealed Resources=none
Internal requirements count=0 size=12
```

**Reply by charleskorn** [reply]
Sorry this is so painful - I don't (yet 🤞) have a Apple Silicon device to test this out for myself. One other thing to check: if you go into Security and Privacy in System Preferences, what option is selected under "Allow apps downloaded from"?

**Reply by smozely** [reply]
Sorry for delay, been having a few days off. The "App Store and identified developers" selection is turned on.

Totally appreciate the effort your putting in! must be a nightmare trying to resolve it without the problematic gear

**Reply by jgv115** [reply]
> @smozely, could you try running the following on your M1 Mac and then try to run Batect again? If so, I might be able to add a workaround for this issue until `jffi` fixes the issue for good:
> 
> ```shell
> cd ~/.batect/cache/0.74.0/ # Replace 0.74.0 with whatever version you're using
> jar xf batect-0.74.0.jar jni/Darwin/libjffi-1.2.jnilib
> codesign -s - ./jni/Darwin/libjffi-1.2.jnilib
> jar uf batect-0.74.0.jar jni/Darwin/libjffi-1.2.jnilib
> ```
> 
> Note that this requires that you at least have the Xcode command line tools installed - if you have Xcode installed, or are using Homebrew, you should already have these available.

This actually worked for me (using 0.76.1 though). I had to change the checksum of the `batect` script file in my project to the newly packed `.jar` but after that things seem to work normally!

**Reply by charleskorn** [reply]
Out of interest @jgv115, what happens if you run `codesign -dv ./jni/Darwin/libjffi-1.2.jnilib` in `~/.batect/cache/0.76.1`?

Also, if you share that `batect-0.76.1.jar` with a colleague running on a M1 Mac, does it work for them as well without them needing to re-sign it?

**Reply by jgv115** [reply]
This is what I get:

```
~/.batect/cache/0.76.1 >> codesign -dv ./jni/Darwin/libjffi-1.2.jnilib
./jni/Darwin/libjffi-1.2.jnilib: code object is not signed at all
```

Unfortunately I don't have another M1 Mac around me - I only have my own. I could upload my .jar here if that helps?

**Reply by charleskorn** [reply]
```
./jni/Darwin/libjffi-1.2.jnilib: code object is not signed at all
```

🤨 That seems wrong... if it's not signed, it shouldn't be able to be loaded.

Could you post the output of `java -version` @jgv115 and @smozely? I'm wondering if the JVM is a contributor to the problem as well.

**Reply by jgv115** [reply]
Ah sorry @charleskorn, I made a mistake and ran it on a fresh batect install. Here is the output on the one I manually signed to get it to work on my MBP:

```
~/.batect/cache/0.76.1 > codesign -dv ./jni/Darwin/libjffi-1.2.jnilib                                 
Executable=/Users/jgv115/.batect/cache/0.76.1/jni/Darwin/libjffi-1.2.jnilib
Identifier=libjffi-1-55554944ddb5c3e598b5378987a6ec9eda8a36da
Format=Mach-O universal (x86_64 arm64)
CodeDirectory v=20400 size=1547 flags=0x2(adhoc) hashes=42+2 location=embedded
Signature=adhoc
Info.plist=not bound
TeamIdentifier=not set
Sealed Resources=none
Internal requirements count=0 size=12
```

For `java -version` I get:

```
openjdk version "17.0.1" 2021-10-19
OpenJDK Runtime Environment Homebrew (build 17.0.1+0)
OpenJDK 64-Bit Server VM Homebrew (build 17.0.1+0, mixed mode, sharing)
```

I hope this helps

**Reply by charleskorn** [reply]
That's the output I'd expect, and it matches what @smozely posted earlier, so I'm intrigued why it isn't working for @smozely...

@jgv115, are you able to ask a colleague (even if they're running on an Intel Mac) to run [the code signing steps you performed](https://github.com/batect/batect/discussions/963#discussioncomment-1528820) on an unmodified version of the `.jar` and then try running that on your M1 machine? I'm trying to understand whether simply having any signature is sufficient, or whether it needs to be a signature from a paid Apple Developer Program certificate.

**Reply by smozely** [reply]
Oh ... I was sure I'd already tried different JVMs.

But yes it wasn't working with this ...
```
openjdk version "1.8.0_302"
OpenJDK Runtime Environment (Zulu 8.56.0.23-CA-macos-aarch64) (build 1.8.0_302-b08)
OpenJDK 64-Bit Server VM (Zulu 8.56.0.23-CA-macos-aarch64) (build 25.302-b08, mixed mode)
```

And now I've tried this version again (just it was the other one I already had on my machine)
```
openjdk version "16.0.1" 2021-04-20
OpenJDK Runtime Environment Zulu16.30+19-CA (build 16.0.1+9)
OpenJDK 64-Bit Server VM Zulu16.30+19-CA (build 16.0.1+9, mixed mode)
```
It seems to work. So maybe yeah just a recent java version is required (I have 8 installed because an old codebase I still maintain, but not the one I'm using batect for)

**Reply by jgv115** [reply]
@smozely is your JDK8 compiled for arm64 or is it running under Rosetta 2?

**Reply by jgv115** [reply]
@charleskorn I signed an unmodified version of the `.jar` on my Intel MacBook Pro, copied this `.jar` over to my M1 Max MacBook Pro. I then replaced the batect script with the SHA-256 checksum of the `.jar` from my Intel machine.

This seems to work, no errors at all!

**Reply by smozely** [reply]
@jgv115 Compiled for ARM I believe

**Reply by charleskorn** [reply]
There have been a bunch of improvements for macOS ARM support in more recent versions of Java - if you can use a more recent version, I'd definitely encourage that.

**Reply by charleskorn** [reply]
Thanks for all your help @smozely and @jgv115 - I should be able to bake in that code signature as a workaround until `jffi` has released an updated version.

**Reply by charleskorn** [reply]
(I've raised https://github.com/jnr/jffi/pull/116 to get this fixed in the library Batect uses as well.)

**Reply by charleskorn** [reply]
@smozely and @jgv115, could you please drop the `.jar` from https://github.com/batect/batect/suites/4419146264/artifacts/116858088 in as a replacement for your existing `.jar` and let me know if that works for you? (The checksum should be `0136bb2019e58d17bf0a077e3f7f730a21ecbb75fd80cdd8ae19940ccf661df2`.)

If that works, I can release that, which should solve the problem for everyone 🎉

**Reply by smozely** [reply]
@charleskorn works for me!

**Reply by charleskorn** [reply]
Awesome, I'll release that now then.

**Comment by charleskorn** [comment, marked as answer]
For anyone coming across this thread: this has been resolved in [0.77.1](https://github.com/batect/batect/releases/tag/0.77.1). You'll need to use a recent version of Java (I recommend the latest long-term support release, version 17).

**Reply by elvay1** [reply]
It still fails on my M1 and java 11.

```
openjdk 11.0.13 2021-10-19 LTS
OpenJDK Runtime Environment (build 11.0.13+8-LTS)
OpenJDK 64-Bit Server VM (build 11.0.13+8-LTS, mixed mode)
```

batect freeze at launch, and no log anywhere to explain the error.

**Reply by charleskorn** [reply]
Hi @elvay1, could you confirm what version of Batect you're using? (Run `./batect --version` and paste the output here.)

I know of one issue where Batect might crash on M1 Macs, but there should be some output to the console when this happens.

**Reply by elvay1** [reply]
Hi @charleskorn ,

Saddly, ./batect --version behaves like i explained on my first post, it freeze, and no output. But the script downloads the jar batect-0.77.1.jar with shasum f9a763b832e8ac8c59334c3f4a752e40afce0d3a

my jdk is liberica-11.0.13+8 if it can help.

Thank you very much!

**Reply by charleskorn** [reply]
OK, try 0.77.2, which I've just released - this contains a fix for the issue I mentioned earlier and may fix your problem.

If not, please try upgrading your JVM to v17.

**Reply by elvay1** [reply]
It's a success, version 0.77.2 fixed my issue, thank you very much!
